### ISSUE TITLE
ModuleNotFoundError: No module named 'typing.io'; 'typing' is not a package

### ISSUE THREAD

**Issue by adamgranthendry** [Almenon/AREPL-vscode#416]
**Bug**

OS: Windows 10 Professional x64, Build 1909
Python: 3.8.10
VSCode: 1.58.2
AREPL: 2.0.2

When using `importlib.resources`, AREPL generates the error:

```
File "C:\Program Files\Python38\lib\importlib\resources.py", line 13, in <module>
    from typing.io import BinaryIO, TextIO
ModuleNotFoundError: No module named 'typing.io'; 'typing' is not a package
```

**Minimal Example**
1. Create `python 3.8.10` virtual environment with `venv` in project folder:
```bash
py -m venv .venv
```
2. Create VSCode user settings to point AREPL to virtual environment:
```json
"AREPL.pythonPath": "${workspaceFolder}\\.venv\\Scripts\\python.exe"
```
3. Create file `demo.py` in project folder
4. Import `importlib.resources` in `demo.py`:
```python
import importlib.resources as rsrc
```
5. Click cat button for AREPL to run repl 

**Expected Behavior**
AREPL uses virtual environment and is able to import `importlib.resources`

**Screenshots**
![image](https://user-images.githubusercontent.com/77850046/126852241-43f3618d-98e2-4791-90e0-c4cf2d062a48.png)

![image](https://user-images.githubusercontent.com/77850046/126852247-ce91ff6b-5d84-4bcf-9cf3-1ac494b9969d.png)

**Comment by Almenon** [comment]
good find, this is indeed a bug :(

**Comment by adamgranthendry** [comment]
@Almenon Thank you! Are there any next steps towards fixing this bug?

**Comment by Almenon** [comment]
I can do some investigation if I had time for it, which I don't. Unless you're interested in taking a look it's likely going to be a while before it will be fixed. If you are interested I can help you out or review a PR.

**Comment by gulash007** [comment]
If I may, this seems to have been a problem with the resources.py file itself. I fixed it for myself by replacing _from typing.io_ with just _from typing_. Don't know why _typing.io_ was even there when the classes were in _typing_. 🤷

**Comment by OrhanYaz** [comment]
> If I may, this seems to have been a problem with the resources.py file itself. I fixed it for myself by replacing _from typing.io_ with just _from typing_. Don't know why _typing.io_ was even there when the classes were in _typing_. 🤷
> 
> Edit: Now I noticed that there is a wrapper class _io_ for the _TextIO_ and _BinaryIO_ classes. Don't know why it doesn't work though, but the fix works for me.

Hello, can you please explain how exactly did you fixed it ? Do i need to do it on VS code. Where can i find the ressource.py file ?

**Comment by gulash007** [comment]
> Hello, can you please explain how exactly did you fixed it ? Do i need to do it on VS code. Where can i find the ressource.py file ?

Sorry for a late reply. For what it's worth, the resources.py file is a part of the importlib library. On my mac it's located at /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/importlib/resources.py.

**Comment by ggonzalez-elec** [comment]
> > Hello, can you please explain how exactly did you fixed it ? Do i need to do it on VS code. Where can i find the ressource.py file ?
> 
> Sorry for a late reply. For what it's worth, the resources.py file is a part of the importlib library. On my mac it's located at /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/importlib/resources.py.

Hi! For those using Windows, you can find resources.py file into the users/<your_user>/AppData/Local/Programs/Python/Python310/Lib/importlib/resources.py. As @gulash007 said, you just need to delete ".io" from "typing.io" and that's sould work.

**Comment by adam-grant-hendry** [comment]
> Don't know why typing.io was even there when the classes were in typing. 🤷

@gulash007 `typing.io` was deprecated in Python 3.8 (see [the `typing` module docs](https://docs.python.org/3/library/typing.html#other-concrete-types):

![image](https://user-images.githubusercontent.com/59346180/195657798-e62d4496-d4a8-4646-9540-2fe0b4742da4.png)

However, this wasn't removed from `importlib` (see [the cpython source](https://github.com/python/cpython/blob/3.8/Lib/importlib/resources.py)). It isn't scheduled to be removed completely until Python 3.12.

`io` is [made into a namespace](https://github.com/python/cpython/blob/0a4f650347fdcfd82d094ab2134ca89584f4e877/Lib/typing.py#L1986) within the `typing` module, but `typing` is not a package (i.e. a folder with a `__init__.py`). One can import `typing.io` perfectly fine in a standard REPL.

The above leads me to believe the error is coming from the TypeScript modules concerning the python environment as used in AREPL.

**Comment by adam-grant-hendry** [comment]
> Hi! For those using Windows, you can find resources.py file into the /users/<your_user>/AppData/Local/Programs/Python/Python310/Lib/importlib/resources.py

@ggonzalez-elec The bug isn't with `resources.py`, but with `AREPL`, so I don't recommend changing `resources.py` because the problem is deeper than that. The change is fine since Python explicitly states one should import `typing` instead of `typing.io`, but it won't be removed until Python 3.12. (Also, every time you install a virtual environment or Python on a new system, you'll have to do this manual change, which isn't a maintainable solution).

Also be careful: you have a local system install of Python 3.10 and are not using a virtual environment, which is why your path is what it is, but this is not what all users will have. `importlib.resources` is a standard library module, so it will always exist in

`Python<3.x>\Lib\importlib\resources.py`

so the problem of finding it boils down to where you have installed Python.

On Windows:

`%USERPROFILE%` = `C:\Users\<user_name>`
`%LOCALAPPDATA%` = `%USERPROFILE%\AppData\Local`
`%PROGRAMFILES%` = `C:\Program Files`
`%PROGRAMFILES(X86)%` = `C:\Program Files (x86)`

**Local User Install:** `%LOCALAPPDATA%\Programs\Python\Python<3.x>
**System Install (32-bt):** `%PROGRAMFILES(X86)%\Python<3.x>
**System Install (64-bit):** `%PROGRAMFILES%\Python<3.x>

In a virtual environment (typically named `.venv`), it lives in the folder `.venv\Lib\importlib\resources.py`.

Again, however, I don't recommend changing `resources.py` because that doesn't solve the actual problem and will cause you headache when dealing with multiple environments.

**Comment by adam-grant-hendry** [comment]
> If I may, this seems to have been a problem with the resources.py file itself.

This is not a problem with `resources.py`. `typing.io` is deprecated and [scheduled to be removed](https://github.com/Almenon/AREPL-vscode/issues/416#issuecomment-1277965222), but `typing.io` can be imported and used in all versions of Python up to `3.11` outside of `AREPL` just fine.

The issue is with the manner `AREPL` is loading `typing.io`. `typing.io` is not a package, as [I explained previously](https://github.com/Almenon/AREPL-vscode/issues/416#issuecomment-1277965222), but `AREPL` is trying to interpret it as such.

The actual problem is how `AREPL` is trying to import `typing.io`, which I believe is somehow related to its typescript code.

I **strongly discourage** users from manually changing `resources.py`. Doing so does not solve the actual problem.

**Comment by Almenon** [comment]
> I believe this is related to AREPL's typescript code.

The Typescript code doesn't have much to do with the python code outside of spawning the process and pipe I/O. The issue is more likely to be with the AREPL python code, which does some funky stuff around importing so there is a fresh state each time. https://github.com/Almenon/AREPL-backend/blob/master/python/arepl_python_evaluator.py

One thing that's pretty weird is that I don't get this bug with python 3.9, only 3.8 and 3.10 (haven't checked 3.11). I also can't reproduce the issue in a unit test.
@adam-grant-hendry do you get this issue with 3.9? 

Also, adding "typing.io" to the modules_to_keep variable in `C:\Users\<yourUser>\.vscode\extensions\almenon.arepl-2.0.3\node_modules\arepl-backend\python\arepl_python_evaluator.py` (insert after line 27) fixes the issue on my end. Does it fix the issue for you?

**Comment by ashwin-interfold** [comment]
I hit this problem today and stumbled upon this conversation. 
@Almenon I tried what you suggest above here, 

> Also, adding "typing.io" to the modules_to_keep variable in C:\Users\<yourUser>\.vscode\extensions\almenon.arepl-2.0.3\node_modules\arepl-backend\python\arepl_python_evaluator.py (insert after line 27) fixes the issue on my end.


It seemed to help. My version of python is 3.10.8

**Comment by Almenon** [comment]
This error should be fixed with the latest release of AREPL (2.0.4)

**Comment by nikcoke** [comment]
For me the issue with the typing.io module being deprecated in python 3.12 onwards outlined [here](https://docs.python.org/3.12/library/typing.html#aliases-to-other-concrete-types). I had an issue running spark locally being produced by one of the imports: from typing.io import BinaryIO. What I did is go to the import file and change from which location we import the BinaryIO library and in perticular: from typing import BinaryIO
